src/water_cmd.cpp

```diff
@@ -247,6 +257,14 @@ void MakeWaterKeepingClass(TileIndex tile, Owner o)
 		default: break;
 	}
 
+	/* Restore the water depth from surrounding tiles */
+	uint8 min_water_depth = WATER_DEPTH_MAX + 1;
+	for (Direction dir = DIR_BEGIN; dir < DIR_END; dir++) {
+		const TileIndex dest = tile + TileOffsByDir(dir);
+		if (IsValidTile(dest) && IsTileType(dest, MP_WATER)) min_water_depth = min(min_water_depth, GetWaterDepth(dest));
+	}
+	if (min_water_depth <= WATER_DEPTH_MAX) SetWaterDepth(tile, min_water_depth);
+
 	if (wc != WATER_CLASS_INVALID) CheckForDockingTile(tile);
```

Comment:
```
	if (wc != WATER_CLASS_INVALID) {
		/* Restore the water depth from surrounding tiles */
		uint8 min_water_depth = WATER_DEPTH_MAX + 1;
		for (Direction dir = DIR_BEGIN; dir < DIR_END; dir++) {
			const TileIndex dest = tile + TileOffsByDir(dir);
			if (IsValidTile(dest) && IsTileType(dest, MP_WATER)) min_water_depth = min(min_water_depth, GetWaterDepth(dest));
		}
		if (min_water_depth <= WATER_DEPTH_MAX) SetWaterDepth(tile, min_water_depth);

		CheckForDockingTile(tile);
	}
```

Reply:
Industries near coasts can trigger an assert in SetWaterDepth, due to the surroundings being a coast with depth 0.